packages/plugin-core/src/services/NoteTraitManager.ts

```diff
@@ -0,0 +1,263 @@
+      return this.cmdRegistar.registeredCommands[trait.id];
+    }
+
+    return undefined;
+  }
+
+  dispose() {
+    if (this._watcher) {
+      this._watcher.dispose();
+    }
+  }
+
+  // ^6fjseznl6au4
+  private async setupSavedTraitsFromFS() {
+    const { wsRoot } = ExtensionProvider.getDWorkspace();
+
+    const userTraitsPath = wsRoot
+      ? path.join(wsRoot, CONSTANTS.DENDRON_USER_NOTE_TRAITS_BASE)
+      : undefined;
+
+    if (userTraitsPath && fs.pathExistsSync(userTraitsPath)) {
+      const files = fs.readdirSync(userTraitsPath);
+
+      // Track some info about how many and what kind of traits users have
+      let traitJSFileCount = 0;
+      let traitInitializedCount = 0;
+      let traitHasSetTitleImplCount = 0;
+      let traitHasSetNameModifierImplCount = 0;
+      let traitHasSetTemplateImplCount = 0;
+
+      files.forEach((file) => {
+        if (file.endsWith(".js")) {
+          traitJSFileCount += 1;
+
+          const resp = this.setupTraitFromJSFile(
+            path.join(userTraitsPath, file)
+          );
+
+          // Don't log an error at this point since we're just initializing - if
+          // a user has some old trait files with errors in their workspace,
+          // don't warn about trait errors until they try to actually use it.
+          if (!ResponseUtil.hasError(resp)) {
+            traitInitializedCount += 1;
+
+            const newNoteTrait = resp.data;
+
+            if (newNoteTrait?.OnCreate?.setTitle) {
+              traitHasSetTitleImplCount += 1;
+            }
+
+            if (newNoteTrait?.OnCreate?.setTemplate) {
+              traitHasSetTemplateImplCount += 1;
+            }
+
+            if (newNoteTrait?.OnWillCreate?.setNameModifier) {
+              traitHasSetNameModifierImplCount += 1;
+            }
+          }
+        }
+      });
+
+      if (traitJSFileCount > 0) {
+        AnalyticsUtils.track(VSCodeEvents.NoteTraitsInitialized, {
+          traitJSFileCount,
+          traitInitializedCount,
+          traitHasSetTitleImplCount,
+          traitHasSetNameModifierImplCount,
+          traitHasSetTemplateImplCount,
+        });
+      }
+    }
+  }
+
+  private setupTraitFromJSFile(fsPath: string): RespV2<NoteTrait> {
+    const traitId = path.basename(fsPath, ".js");
+
+    this.L.info("Registering User Defined Note Trait with ID " + traitId);
+    const newNoteTrait = new UserDefinedTraitV1(traitId, fsPath);
+
+    const resp = this.registerTrait(newNoteTrait);
+
+    return {
+      data: newNoteTrait,
+      error: resp.error,
+    };
+  }
+
+  private async setupFileWatcherForTraitFileChanges() {
+    const userTraitsPath = this._wsRoot
+      ? path.join(this._wsRoot, CONSTANTS.DENDRON_USER_NOTE_TRAITS_BASE)
+      : undefined;
+
+    if (!userTraitsPath) {
+      return;
+    }
+
+    const pattern = new vscode.RelativePattern(userTraitsPath, "*.js");
+
+    this._watcher = vscode.workspace.createFileSystemWatcher(
```

Comment:
will this work for native workspaces or only code workspaces?

Reply:
Native workspaces will include the traits path in the folder set, so there's no problem there.  There is however an issue with self-contained vaults, which seems to exclude all paths that aren't 'notes' in the generated code workspace file.  So pods, traits, etc. don't appear - I'm not sure if that was intentional or not, but I can look into it in the next iteration.